src/h3d/col/ConvexHull.hx

```diff
@@ -0,0 +1,341 @@
+
+	// Mapping points back to faces
+	public var pointFaces: Array<Array<Int>> = new Array();
+
+	// Directional support function start points
+	var suppPX: Int;
+	var suppNX: Int;
+	var suppPY: Int;
+	var suppNY: Int;
+	var suppPZ: Int;
+	var suppNZ: Int;
+
+	public function new(points, faces) {
+		this.points = points;
+		this.faces = faces;
+
+		calcPointFaces();
+		calcEdges();
+		calcFaceConnections();
+		calcPlanes();
+		calcSuppPoints();
+
+		if (VALIDATE)
+			validate();
+	}
+
+	function calcPointFaces() {
+		for (p in 0...points.length) {
+			pointFaces[p] = [];
+		}
+
+		for (f in 0...Std.int(faces.length/3)) {
+			for (pi in 0...3) {
+				var pInd = facePoint(f, pi);
+				pointFaces[pInd].push(f);
+			}
+		}
+	}
+
+	function calcEdges() {
+		for (pInd in 0...points.length) {
+			var connList = new Array();
+			pointConnections[pInd] = connList;
+			var dirList = new Array();
+			edgeDirs[pInd] = dirList;
+			var faceList = pointFaces[pInd];
+			for (f in faceList) {
+				var pOther = -1;
+				for (pi in 0...3) {
+					if (facePoint(f, pi) == pInd) {
+						pOther = facePointMod(f, pi+1);
+						break;
+					}
+				}
+				Debug.assert(pOther != -1);
+
+				connList.push(pOther);
+				var edgeDir = points[pOther].sub(points[pInd]);
+				edgeDir.normalize();
+				dirList.push(edgeDir);
+			}
+
+		}
+	}
+
+	function calcFaceConnections() {
+		for (f in 0...Std.int(faces.length/3)) {
+			for (pi in 0...3) {
+				var p0 = facePoint(f, pi);
+				var p1 = facePointMod(f, pi+1);
+				// Due to the constraints, there must be a face
+				// with points p1, p0, pX or some rotation of this list
+				// Go to p1 and find the face that satisfies this constraint
+				var fInd = -1;
+				for (testFace in pointFaces[p0]) {
+					var testP = finishFace(testFace, p1, p0);
+					if (testP != -1) {
+						fInd = testFace;
+						break;
+					}
+				}
+				Debug.assert(fInd != -1);
+				faceConnections.push(fInd);
+			}
+		}
+	}
+	function calcPlanes() {
+		for (f in 0...Std.int(faces.length/3)) {
+			var p0 = points[facePoint(f,0)];
+			var p1 = points[facePoint(f,1)];
+			var p2 = points[facePoint(f,2)];
+			var plane = Plane.fromPoints(p0.toPoint(), p1.toPoint(), p2.toPoint());
+			plane.normalize();
+			facePlanes.push(plane);
+		}
+	}
+
+	function support(dir: Vector): Vector {
+		var p = -1;
+		// Choose a starting point based on the 6 cardinal directions
+		if (dir.x > 0.0 && dir.x > dir.y && dir.x > dir.z) {
+			p = supportImp(dir, suppPX);
+		} else if (dir.x < 0.0 && dir.x < dir.y && dir.x < dir.z) {
+			p = supportImp(dir, suppNX);
+		} else if (dir.y > 0 && dir.y > dir.z) {
+			p = supportImp(dir, suppPY);
+		} else if (dir.y < 0 && dir.y < dir.z) {
+			p = supportImp(dir, suppNY);
+		} else if(dir.z > 0) {
+			p = supportImp(dir, suppPZ);
+		} else {
+			p = supportImp(dir, suppNZ);
+		}
+
+		return points[p];
+	}
+
+	function supportImp(dir: Vector, startPoint: Int): Int {
+		var curP = startPoint;
+		while(true) {
```

Comment:
Curious to hear your thoughts. But instead of infinite loops I like to use an arbitrarily large counter instead, so termination is guaranteed.

Reply:
Yeah that is safer. In this case it's kind of impossible since points are finite and you can't go backwards. At least short of a logic error or serious floating point error.

On the flip side I find "while(true)" has better semantic SNR than guarded loops.

Reply:
Ooops, that wasn't supposed to be a review. I clicked the big green button.

Reply:
Bad UI

Reply:
You can't review the review.

Reply:
2/5 delivery guy ate a slice of my pizza.